lib/unlock-helpers.js

```diff
@@ -0,0 +1,221 @@
+import logger from './logger';
+import { sleep } from 'asyncbox';
+
+const PIN_UNLOCK = "pin";
+const PASSWORD_UNLOCK = "password";
+const PATTERN_UNLOCK = "pattern";
+const UNLOCK_TYPES = [PIN_UNLOCK, PASSWORD_UNLOCK, PATTERN_UNLOCK];
+const KEYCODE_NUMPAD_ENTER = "66";
+const UNLOCK_WAIT_TIME = 100;
+const HIDE_KEYBOARD_WAIT_TIME = 100;
+const INPUT_KEYS_WAIT_TIME = 100;
+
+let helpers = {};
+helpers.isValidUnlockType = function (type) {
+  return UNLOCK_TYPES.indexOf(type) !== -1;
+};
+
+helpers.isValidKey = function (type, key) {
+  if (typeof key === 'undefined') {
+    return false;
+  }
+  if (type === PIN_UNLOCK) {
+    return /^[0-9]+$/.test(key.trim());
+  }
+  if (type === PATTERN_UNLOCK) {
+    if (!/^[1-9]+$/.test(key.trim())) {
+      return false;
+    }
+    return !(/([1-9]).*?\1/.test(key.trim()));
+  }
+  // Dont trim password key, you can use blank spaces in your android password
+  // ¯\_(ツ)_/¯
+  if (type === PASSWORD_UNLOCK) {
+    return /.{4,}/g.test(key);
+  }
+  throw new Error(`Invalid unlock type ${type}`);
+};
+
+helpers.dismissKeyguard = async function (driver, adb) {
+  let isKeyboardShown = await driver.isKeyboardShown();
+  if (isKeyboardShown) {
+    await driver.hideKeyboard();
+    // Waits a bit for the keyboard to hide
+    await sleep(HIDE_KEYBOARD_WAIT_TIME);
+  }
+  // dismiss notifications
+  logger.info("Dismiss notifications from unlock view");
+  await adb.shell(["service", "call", "notification", "1"]);
+  await adb.back();
+  if (await adb.getApiLevel() > 21) {
+    logger.info("Trying to dismiss keyguard");
+    await adb.shell(["wm", "dismiss-keyguard"]);
+    return;
+  }
+  logger.info("Swiping up to dismiss keyguard");
+  await helpers.swipeUp(driver);
+};
+
+helpers.swipeUp = async function (driver) {
+  let windowSize = await driver.getWindowSize();
+  let x0 = parseInt(windowSize.x / 2, 10);
+  let y0 = windowSize.y - 10;
+  let yP = 100;
+  let actions = [
+    {action: 'press', options: {element: null, x: x0, y: y0}},
+    {action: 'moveTo', options: {element: null, x: x0, y: yP}},
+    {action: 'release'}
+  ];
+  await driver.performTouch(actions);
+};
+
+helpers.encodePassword = function (key) {
+  return key.replace(/\s/ig, "%s");
+};
+
+helpers.stringKeyToArr = function (key) {
+  return key.trim().replace(/\s+/, '').split(/\s*/);
```

Comment:
```
.replace(/\s+/, '')
```

will replace only the very first occurrence and

```
.replace(/\s+/g, '')
```

will replace all occurrences